numpy/lib/index_tricks.py

```diff
@@ -589,6 +590,336 @@ def __iter__(self):
     next = __next__
 
 
+def fproduct(*iterables, **kwargs):
+    """
+    Like itertools.product, but in fortran order
+    """
+    return (i[::-1] for i in itertools.product(*iterables[::-1], **kwargs))
+
+
+class ndrange(_nx.collections_abc.Sequence):
+    """
+    An N-dimensional range object that returns an iterator (or reversed
+    iterator) tthat can produce a sequence of tuples from
+    start (inclusive tuple) to stop (exclusive tuple) by step.
+
+    start defaults to a tuple of zeros and stop is omited!
+
+    When step is given, it specifies the increment (or decrement).
+
+    ndrange is designed to behave much like the Python3 ``range`` object.
+    See `python class range documentation
+    <https://docs.python.org/3/library/stdtypes.html#range>`_ for more details.
+    About valid parameter types and behaviour peculiarities. One notable
+    difference is that the the iterator returned by ``ndrange`` will always
+    return tuples as opposed to singletons.
+
+    Unlike ``ndindex``, ``ndrange`` is not an iterator.
+    You should prefer this ``ndrange`` object over ``ndindex`` as ``ndrange``
+    allows for multi-dimensional slicing.
+
+    Parameters
+    ----------
+    stop: tuple of ints (or single int)
+        Specifies the stopping tuple index.
+    start: tuple of ints (or single int)
+        If provided, must be the same length as stop. Specifies the starting
+        tuple index.
+    step: tuple of ints (or single int)
+        If provided, must be the same length as stop. Specifies the increment
+        along each axis.
+
+
+    See Also
+    --------
+    flatiter, ndenumerate, ndindex
+
+    Notes
+    -----
+    ndrange is not designed as a substitute for vectorized numeric code.
+    Whenever possible, you should use vectorized routines for mathematical
+    functions. ndrange is provided primarily for convenience for iteration over
+    multi-dimensional collections where the operations performed on the
+    collections are the bottlekneck.
+
+    In Python 2.7, this object is based on multiple ``range`` objects, lists, as
+    opposed to ``xrange`` objects.
+
+    Examples
+    --------
+    This is likely the most natural way to iterate through an ndrange object.
+    Here, the first parameter is the shape of the array ``shape`` attribute
+    of an array and we slice through the ``ndrange`` object.
+
+    >>> a = np.zeros((3, 3), dtype='object')
+    >>> # [...] # operate on ``a``
+    >>> for i in np.ndrange(a.shape)[::2]:
```

Comment:
I would kind of like an example with multi-dim index, like in your mailing list post. I liked that demonstration of the usefulness of ndrange. 

Maybe, just change this example to do `for i in np.ndrange(a.shape)[:, ::2]:` ? Or add an extra example along those lines.

Reply:
You are right, I modified the example. It is likely people looking for this are going to be looking for this slicing operation.